File: wheel.rb
require_relative 'pqueue'

# Not quite a timing wheel https://blog.acolyer.org/2015/11/23/hashed-and-hierarchical-timing-wheels/
# But a similar idea: hash of { priority => list_of_items },
# then a separate priority queue with list of priorities, so we can have dynamic step sizes.
# Optimized for multiple items with each priority, but large gaps between priorities.
# If you squint, that describes a scheduler. Thus "timing wheel".
class Wheel
  Bucket = Data.define(:priority, :list)

  def initialize
    @q = PQueue.new() { |a,b| a.priority <=> b.priority }
    @buckets = {}
  end

  def push(item, priority)
    bucket = @buckets[priority]
    unless bucket
      bucket = Bucket[priority, []]
      @buckets[priority] = bucket
      @q << bucket
    end
    bucket.list << item
  end

  def pop
    loop do
      bucket = @q.peek
      return nil unless bucket
      if bucket.list.empty?
        @q.pop
        @buckets.delete bucket.priority
        next
      end
      return bucket.list.pop
    end
  end
end

if __FILE__ == $0
  require_relative 'asserts'
  q = Wheel.new
  [5, 3, 9, 6, 4].each { q.push _1, _1 }
  assert_eq q.pop, 3
  assert_eq q.pop, 4
  assert_eq q.pop, 5
  assert_eq q.pop, 6
  q.push 11, 11
  q.push 3, 3
  q.push 3, 3
  assert_eq q.pop, 3
  assert_eq q.pop, 3
  assert_eq q.pop, 9
  assert_eq q.pop, 11
  assert_eq q.pop, nil
  puts "tests passed"
end

File: other/wheel/wheel.go
package wheel

import "github.com/codekitchen/advent-of-code/other/pqueue"

type Wheel[T any] struct {
	q       pqueue.PQueue[struct{}]
	buckets map[int64]*bucket[T]
}

type bucket[T any] []T

func (w *Wheel[T]) Push(value T, priority int64) {
	w.init()
	b, ok := w.buckets[priority]
	if !ok {
		b = &bucket[T]{}
		w.buckets[priority] = b
		w.q.Push(struct{}{}, priority)
	}
	*b = append(*b, value)
}

func (w *Wheel[T]) Pop() (T, int64, bool) {
	var zero T
	for {
		_, priority, ok := w.q.Peek()
		if !ok {
			return zero, 0, false
		}
		b := w.buckets[priority]
		if len(*b) == 0 {
			delete(w.buckets, priority)
			w.q.Pop()
			continue
		}
		return b.pop(), priority, true
	}
}

func (b *bucket[T]) pop() T {
	var zero T
	i := len(*b) - 1
	ret := (*b)[i]
	(*b)[i] = zero
	*b = (*b)[:i]
	return ret
}

func (w *Wheel[T]) init() {
	if w.buckets == nil {
		w.buckets = make(map[int64]*bucket[T])
	}
}
